File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/studio/Studio.java
package org.taonaw.studio_reservation.domain.model.studio;

import lombok.NonNull;
import org.taonaw.studio_reservation.domain.shared.exception.ErrorNotification;

import java.util.Objects;

public class Studio {
    private StudioId id;
    private StudioName name;
    private StartTime startTime;
    private EquipmentMaxUsableCounts equipmentMaxUsableCounts = EquipmentMaxUsableCounts.empty();

    private Studio(@NonNull StudioId id) {
        this.id = id;
    }

    public static Studio create(
            @NonNull StudioName name,
            @NonNull StartTime startTime,
            @NonNull EquipmentMaxUsableCounts equipmentMaxUsableCounts) {

        var errorNotification = new ErrorNotification();
        errorNotification.addError(equipmentMaxUsableCounts.validateDuplicated());
        errorNotification.throwIfHasErrors("スタジオ内容に不備があります。");

        var instance = new Studio(StudioId.newId());
        instance.name = name;
        instance.startTime = startTime;
        instance.equipmentMaxUsableCounts = equipmentMaxUsableCounts;
        return instance;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Studio studio = (Studio) o;
        return id.equals(studio.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/backend/facility/src/main/kotlin/org/taonaw/facility/domain/model/studio/Studio.kt
package org.taonaw.facility.domain.model.studio

class Studio private constructor(
        val studioId: StudioId,
        val studioName: StudioName,
        val studioSize: StudioSize,
        val startTime: StartTime,
        val studioUsableStatus: StudioUsableStatus,
        val maxRentalEquipmentQuantities: MaxRentalEquipmentQuantities) {

    companion object {
        fun create(
                studioName: StudioName,
                studioSize: StudioSize,
                startTime: StartTime,
                maxRentalEquipmentQuantities: MaxRentalEquipmentQuantities): Studio {

            maxRentalEquipmentQuantities.validateDuplicated()?.throwErr()

            return Studio(StudioId.newId(), studioName, studioSize, startTime, StudioUsableStatus.SUSPENDED, maxRentalEquipmentQuantities)
        }
    }

    fun change(
            studioName: StudioName,
            studioSize: StudioSize,
            startTime: StartTime,
            studioUsableStatus: StudioUsableStatus,
            maxRentalEquipmentQuantities: MaxRentalEquipmentQuantities): Studio {

        maxRentalEquipmentQuantities.validateDuplicated()?.throwErr()

        return Studio(studioId, studioName, studioSize, startTime, studioUsableStatus, maxRentalEquipmentQuantities)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Studio

        if (studioId != other.studioId) return false

        return true
    }

    override fun hashCode(): Int {
        return studioId.hashCode()
    }
}